### DISCUSSION TITLE
Update EVM Version to Shanghai

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/evm-monorepo#925] [General]
I recommend that we should update the [`evm_version`](https://github.com/sablier-labs/v2-core/blob/8d3ab20abd0c0570313a4b4af53315ac0e5398e2/foundry.toml#L4) to `shanghai` to take advantage of the `PUSH0` opcode. This evm version is supported by all chains we deploy to.

## Context

Currently, we rely on `paris` version of evm for deployment. Foundry uses the value of `evm_version` to generate bytecode during on-chain deployments.

Since **paris**, there have been two mainnet hardforks - **shanghai** and **dencun**. While **dencun** did not introduce anything relevant to us, **shanghai** introduced `PUSH0` - an opcode to place "0" on the stack.

Benefits of `PUSH0`:
1. `PUSH0` consumes 1 less gas compared to `PUSH1 0` making deployed bytecode smaller and reducing deployment costs.
2. Cheaper transactions for functions that require placing "0" on the stack.

## Support for `PUSH0` Opcode Across Chains

I have deployed the [following bytecode](https://bytegraph.xyz/bytecode/7ec82be8ef4d997665416023f37c4c4a/graph) on every chain supported by Sablier (except Lightlink, I didn't deploy on it). All the deployments were successful which means all of them support `PUSH0`.

```bash
0x6080604052348015600e575f80fd5b50600436106026575f3560e01c80632ee921d714602a575b5f80fd5b5f60405190815260200160405180910390f3fea2646970667358221220721ae7de3b3c2a6157c6cb0c3bd4239cf5cd317e94590094fcee9bde551aa99464736f6c63430008170033
```

> [!IMPORTANT]  
> Since zksync has its own variation of `foundry`, the deployed bytecode on zksync is also different. Therefore, updating the evm version to **shanghai** will have no impact on zksync deployment.

RFC @sablier-labs/solidity.

**Comment by PaulRBerg** [comment]
Thanks for the detailed investigation!

Happy to switch to Shanghai. The reason we went with Paris is that when we first deployed V2 in 2023, it wasn't supported by most rollups.

> except Lightlink, I didn't deploy on it

Sorry to be a pain, but could you do that, please?

LightLink is a paid customer, and we've promised to host their protocol for at least a year.

**Reply by smol-ninja** [reply]
Lightlink supports `PUSH0` as well. https://phoenix.lightlink.io/tx/0x3f7bda3036770d75db38e54eaf0a640e94100681b3d687bb824bf33268debe9e

**Reply by PaulRBerg** [reply]
Fantastic, in this case let's proceed with switching to Shanghai.

Note: you might also have to do this in PRBMath and all upstream dependencies.

**Reply by smol-ninja** [reply]
It's not required for upstream dependencies since this change only relates to the build process. When we compile `v2-core`, Foundry handles the compilation of all dependencies. Regardless, I will update it in `PRBMath` too.

**Reply by 40404040404040l** [reply]
Fantastic, in this case let's proceed with switching to Shanghai or Dubai.

**Reply by LegalM0ney** [reply]
You can have very good experience with zksync

**Comment by smol-ninja** [comment]
Issues:
- https://github.com/sablier-labs/v2-core/issues/926
- https://github.com/sablier-labs/v2-periphery/issues/347
